src/main/java/com/rocketden/main/service/GameManagementService.java

```diff
@@ -154,4 +164,26 @@ public void updateCode(String roomId, String userId, PlayerCode playerCode) {
         liveGameService.updateCode(game.getPlayers().get(userId), playerCode);
     }
 
+    /**
+     * If all players have solved the problem, update the game and send
+     * socket message indicating as such.
+     * (Depending on the game setting, this may or may not end the game).
+     */
+    public void conditionalSolvedSocketMessage(Game game) {
```

Comment:
On second thought, what do you think about moving this method to the submit service? I was thinking and after each submission, the submit service already sends a socket update, so this would cause 2 messages to be sent on every all solved game end. Maybe instead after the user code is judged/dummy submission is created, it runs a check for whether to set allSolved to true, and then proceeds as normal

Might also help to keep GameManagementService less cluttered

Reply:
I think it makes sense to move this method to submit service -- done! And yeah, that also makes sense not to actually send the socket message through this method, since it will be sent through the latter socket message -- I'll update this to reflect those changes.

Reply:
Just to clarify a little bit - I was thinking the call to the conditionalSolvedSocketMessage (which could probably be renamed now that it's just checking allSolved) could happen perhaps at submit service line 44 (after submission is graded/determined to be success but before socket is sent) and then the method proceeds to send the socket update and return submission as before (meaning game management service 132-145 would no longer be necessary) - mainly to make submitService contain all submission-related logic and to keep the game management service only responsible for checking API errors beforehand and delegating responsibility to the correct class

This might also be related to the other two open comments: 

> Not in our current setup -- the situation in which it would be useful is if the game did not end when everybody solves the problem, but rather when the time runs out if players are competing to make the best solution (fastest or least memory). Do you think I should keep this in for that reason, or no?

After the recent changes would this still be necessary? Assuming we continue the pattern of sending a socket update after each user submission, then even if everyone solves the problem we'd still be sending/expecting a socket update after another person submits

> 1. Yeah this is true, this is actually used for kind of a hacky purpose: in the test of this method in GameManagementServiceTests.java, the submissionDto actually is null if you don't mock it, so I used this check to prevent an error from occurring. I'm happy to instead mock the submissionDto return, just let me know.
> 2. That's a good point, because that only changes

For 1, if the call to the conditional allSolved method is moved to the submit service method, would that remove the null mocking issue? Although I'm not sure exactly what the test covers or if this would require additional testing elsewhere to make up for it. For 2, moving that call could at least I guess simplify that if statement, since submit service will have logic/an if statement in the future to determine whether to set player.getSolved to true, and that call to the conditional check could be made at the same place

Reply:
Ah, that makes sense -- OK, I added that code directly into the `submitSolution` method in `SubmitService.java`!